---
title: "Bioinformatics project (2022-2023)"
author: "Samuele Bompani, Luigi Foscari"
date: "May 2023"
output:
  html_notebook:
    toc: yes
    number_sections: yes
    toc_float: yes
    theme: cerulean
    fig_caption: yes
editor_options: 
  markdown: 
    wrap: 72
---		

# Install and load packages

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install(
  c("curatedTCGAData",
    "TCGAutils",
    "TCGAbiolinks",
    "NetPreProc",
    "remotes",
    "Shamir-Lab/NEMO/NEMO",
    "SNFtool",
    "caret",
    "cluster",
    "mclustcomp",
    "tsne"),
  ask = FALSE
);
```

```{r}
library("curatedTCGAData");
library("TCGAbiolinks");
library("TCGAutils");
library("NetPreProc");
library("SNFtool");
library("caret");
library("cluster"); #pam
library("mclustcomp");
library("NEMO");
library("tsne");
```

# Multi-omics data (1)

```{r}
assays <- c("miRNASeqGene", "RNASeq2Gene", "RPPAArray");
mo <- curatedTCGAData(diseaseCode = "PRAD", assays = assays, dry.run = FALSE, version = "2.0.1", verbose = FALSE);
mo <- mo[, , paste0("PRAD", "_", assays, "-20160128")];
mo;
```

# Filtering (2)

```{r warning=FALSE}
# Consider only primary solid tumors:
primary <- TCGAutils::TCGAsampleSelect(colnames(mo), c("01"));
mo <- mo[, primary, ];

# Check for replicates (anyReplicated checks the so called biological or 
# primary unit in the sampleMap of the MultiAssayExperiment object, that
# corresponds to the first 12 characters of the barcodes for TCGA data):
check_rep <- anyReplicated(mo);
print(check_rep);
```

```{r}
# The information regarding if the sample is FFPE is stored in the clinical data,
# which are accessible using colData(). 
no_ffpe <- which(as.data.frame(colData(mo))$patient.samples.sample.is_ffpe == "no");
mo <- mo[, no_ffpe, ];

# Obtain samples having all the considered omics:
complete <- intersectColumns(mo);

# Extract assays in list:
complete <- assays(complete);

# Obtain matrices samples x features:
complete <- lapply(complete, FUN = t)

```

```{r}
# Remove features having NAs (present only in proteomics data):
complete[[3]] <- complete[[3]][, colSums(is.na(complete[[3]])) == 0];

# Remove features with near zero variance and retain top 100 features 
# having higher variance:
nf <- 100;
for(i in 1:length(complete)) {
    idx <- caret::nearZeroVar(complete[[i]])
    message(paste("Removed ", length(idx), "features from", names(complete)[i]));
    if(length(idx) != 0)
        complete[[i]] <- complete[[i]][, -idx];

    if(ncol(complete[[i]]) <= nf) next
    
    vars <- apply(complete[[i]], 2, var);
    idx <- sort(vars, index.return = TRUE, decreasing = TRUE)$ix;
    
    complete[[i]] <- complete[[i]][, idx[1:nf]];
}

# Perform features standardization using z-score:
zscore <- function(data) {
    zscore_vec <- function(x) { return ((x - mean(x)) / sd(x))}
    data <- apply(data, 2, zscore_vec)
    return(data)
}

complete <- lapply(complete, zscore);

# Clean barcodes retaining only "Project-TSS-Participant":
for(v in 1:length(complete))
    rownames(complete[[v]]) <- substr(rownames(complete[[v]]), 1, 12);
```

# Download the desease subtypes (3 - 4)

```{r}
# Download disease subtypes from TCGAbiolinks:
subtypes <- as.data.frame(TCGAbiolinks::PanCancerAtlas_subtypes());
subtypes <- subtypes[subtypes$cancer.type == "PRAD", ];

# Retain only primary solid tumors and select samples in common with omics data
# (in the same order):
subtypes <- subtypes[TCGAutils::TCGAsampleSelect(subtypes$pan.samplesID, "01"), ];

# Retain from the subtypes only information regarding samples in the multi-omic dataset
sub_select <- substr(subtypes$pan.samplesID, 1, 12) %in% rownames(complete[[1]]);
subtypes <- subtypes[sub_select, ];

# Clean row names
rownames(subtypes) <- substr(subtypes$pan.samplesID, 1, 12);

# Remove subtypes without an associate multi-omic sample
subtypes <- subtypes[rownames(complete[[1]]),];

# Remove subtypes without an associated iCluster molecular subtype
subtypes <- subtypes[!is.na(subtypes$Subtype_Integrative), ];

# Retain from the multi-omics dataset only samples having an associated subtype
for (i in 1:length(complete)) complete[[i]] <- complete[[i]][rownames(subtypes), ]

# Print number of samples for each subtype:
table(subtypes$Subtype_Integrative);
```

```{r}
# Compute similarity matrix for each data source using the scaled exponential euclidean distance:
sim <- list();
for(i in 1:length(complete)) {
	# Compute the distance between the rows of the matrix
    dist <- (dist2(as.matrix(complete[[i]]), as.matrix(complete[[i]])))^(1/2);
    
    # Compute the similarity matrix
    sim[[i]] <- affinityMatrix(dist);
        
    # Normalize in L1 norm
    # sim[[i]] <- sweep(sim[[i]], 1, rowSums(sim[[i]]), FUN = "/");
}
```

# Similarity Network Fusion (5)

```{r}
# Integration of multi-omics data using Similarity Network Fusion:
M_snf <- SNF(sim, K = 20, t = 20)
```

# Mean Fusion (6)

```{r}
# Integration of multi-omics data using the average
M_mean <- Reduce("+", sim) / length(sim)
```

# NEMO (7)

```{r}
# Integration of multi-omics data using NEMO
# Constructs a single affinity graph measuring similarity across different omics.
# the given parameter is a list of the data to be clustered, where each an entry
# is a matrix of *features x samples* and k the number of neighbors to use for each omic.

t_complete <- lapply(complete, FUN = t)
M_nemo <- nemo.affinity.graph(t_complete, k = 20)
```

# PAM (8)

```{r}
clusterings = list();
```

```{r}
k <- length(unique(na.omit(subtypes$Subtype_Integrative)));

# OMIC (a)
for (i in 1:3) {
	dist <- 1 - NetPreProc::Prob.norm(sim[[i]]);
	clusterings[assays[[i]]] <- list(pam(dist, k = k, diss = TRUE, keep.diss = TRUE));
}

# MEAN (b)
dist <- 1 - NetPreProc::Prob.norm(M_mean);
clusterings$mean <- pam(dist, k = k, diss = TRUE, keep.diss = TRUE);

# SNF (c)
dist <- 1 - M_snf
clusterings$snf_pam <- pam(dist, k = k, diss = TRUE, keep.diss = TRUE);

# NEMO (d)
dist <- 1 - NetPreProc::Prob.norm(M_nemo);
clusterings$nemo_pam <- pam(dist, k = k, diss = TRUE, keep.diss = TRUE)
```

```{r}
# print(sim[1])
a <- pam(1 - NetPreProc::Prob.norm(M_nemo), k = k, diss = TRUE, keep.diss = TRUE);
s <- silhouette(a)
plot(s, main = "silhouette")
```

# NEMO clustering (9)

```{r}
clusterings$nemo_spectral$clustering <- nemo.clustering(t_complete, num.clusters = k)
```

# Spectral clustering (10)

```{r}
clusterings$snf_spectral$clustering <- SNFtool::spectralClustering(M_snf, K = k)
```

# Graphs (11)

```{r}
labels <- as.numeric(factor(subtypes$Subtype_Integrative, levels = unique(subtypes$Subtype_Integrative)));

types <- c("rand", "adjrand", "nmi1", "jent");
for (type in types) {
	df <- stack(lapply(clusterings, function(c) mclustcomp(c$clustering, labels, types = type)$scores))
	row.names(df) <- df$ind
	df$ind <- NULL
	names(df) <- c(type)
	df <- df[order(df[, type], decreasing = TRUE), , drop = FALSE]
	print(df)
}
```

```{r}
oldp <- par(mar = c(5.1, 8.1, 2.1, 1.1))

data <- matrix(nrow = k, ncol = length(clusterings) + 1)
for (i in 1:length(clusterings))
	data[,i] <- as.vector(table(clusterings[[i]]$clustering))
data[,length(clusterings) + 1] <- as.vector(table(labels))

barplot(data, names.arg = append(names(clusterings), "iCluster"),
  horiz = TRUE, las = 1, col = c("#5D2A42", "#FB6376", "#FCB1A6"),
	main = "Sample distribution in clusters")

par(oldp)
```

```{r}
for (name in names(clusterings)) {
  print(name)
  print(as.data.frame(clusterings[[name]]$clusinfo))
}
```

```{r}
reduction <- tsne(complete[[2]], k=2)
```

```{r}
for (name in names(clusterings))
	plot(reduction, col = clusterings[[name]]$clustering, xlab = "x", ylab = "y", main = name)
plot(reduction, col = labels, xlab = "x", ylab = "y", main = "iCluster")
```
```{r}
pcairis <- princomp(complete[[3]])
for (name in names(clusterings))
	plot(pcairis$scores, col = clusterings[[name]]$clustering, main = name)
plot(pcairis$scores, col = labels, main = "iCluster")
```
